### ISSUE TITLE
eval_ppl.py Bug Report

### ISSUE THREAD

**Issue by elmo1032** [elmo1032/BitDelta#14]
Here are some potential bugs in the code:

1. The code uses `tqdm` to create a progress bar for loading the dataset, but it does not pass the length of the dataset to `tqdm`. This means that the progress bar will not display accurate progress. To fix this, you can pass the length of the dataset to `tqdm` like this:
```
tqdm(dataset, total=len(dataset))
```
2. The code uses the `tokenizer.decode` method to print the first 100 tokens of the encoded text, but it does not specify which tokenizer to use. This will cause a `NameError` if the `tokenizer` variable has not been defined. To fix this, you should define the `tokenizer` variable before using it.
3. The code uses the `size` parameter of the `get_dataset` function to specify the number of evaluation samples to load, but it does not check whether the returned dataset actually contains this many samples. This could cause an `IndexError` if the dataset contains fewer samples than specified. To fix this, you can check the length of the dataset before iterating over it.
4. The code uses the `torch.float16` data type for the model to save memory, but it does not check whether the device supports this data type. This could cause a `RuntimeError` if the device does not support `torch.float16`. To fix this, you can check the `type` attribute of the `device` variable to see whether it supports `torch.float16`.
5. The code uses the `device_map` parameter of the `AutoModelForCausalLM.from_pretrained` method to automatically map model weights to the device, but it does not check whether the device has enough memory to hold the model. This could cause an `OutOfMemoryError` if the device does not have enough memory. To fix this, you can check the `total_memory` attribute of the `device` variable to see whether it has enough memory for the model.
6. The code uses the `low_cpu_mem_usage` parameter of the `AutoModelForCausalLM.from_pretrained` method to reduce memory usage on the CPU, but it does not check whether the model will actually be run on the CPU. If the model is run on the GPU, this parameter will have no effect. To fix this, you can check the `device` variable to see whether the model will be run on the CPU or the GPU.
7. The code uses the `torch.no_grad` context manager to disable gradient computation during evaluation, but it does not check whether the model requires gradients to be enabled. If the model requires gradients to be enabled, this could cause errors during evaluation. To fix this, you can check the `requires_grad` attribute of the model's parameters to see whether they require gradients to be enabled.
8. The code uses the `torch.exp` function to compute the perplexity, but it does not check whether the result is a valid number. If the result is negative infinity (for example, if the loss is very large), the `torch.exp` function will return `nan`. To fix this, you can check whether the result is `nan` before printing it.
9. The code saves the perplexity to a text file using the `write` method, but it does not check whether the perplexity is a string. If the perplexity is a number, this will cause a `TypeError`. To fix this, you can convert the perplexity to a string before writing it to the file.

I hope this helps! Let me know if you have any questions.
